pubsub/gochannel/pubsub.go

```diff
@@ -340,11 +349,18 @@ func (s *subscriber) Close() {
 	close(s.outputChannel)
 }
 
-func (s *subscriber) sendMessageToSubscriber(msg *message.Message, logFields watermill.LogFields) {
+func (s *subscriber) sendMessageToSubscriber(msg *message.Message, logFields watermill.LogFields, preserveContext bool) {
 	s.sending.Lock()
 	defer s.sending.Unlock()
 
-	ctx, cancelCtx := context.WithCancel(s.ctx)
+	var ctx context.Context
+	var cancelCtx context.CancelFunc
+
+	if preserveContext {
+		ctx, cancelCtx = context.WithCancel(msg.Context())
```

Comment:
Can you describe what's the use case here? 🤔 It seems to be a bit more than just preserving context among messages.

Reply:
Excellent find, removed the WithCancel when preserving context.

WithCancel makes sense when we are using the subscriber context, and done is called on the subscriber context itself. 

A quick question though - Even if the subscriber main context is cancelled, that time the messages will still be processed further.

Reply:
Hey @yashb042 after considering it, I think your original idea made more sense: The message's context should also be wrapped with the cancel, otherwise closing the subscriber wouldn't close the message's context. Sorry about the confusion!

Reply:
@m110 Could you please tell me what is the point of keeping cancelCtx here, for both the cases
As I see it, the messages are still being passed to the subscriber irrespective of whether the context was cancelled or not.

Reply:
@yashb042 It's because the context is passed to the message and is likely to be used in the command handler. For example, when executing a database query, you would pass `msg.Context()`. If the subscriber closes, the context gets canceled, which also cancels the database query. 

Let me know if this makes sense?

Reply:
It does make sense

But let me ask something.

Is there a case where there is a need for the context not to be the one of the message ?

I'm asking because the previous code was creating a context. And now we add an option to preserve it.

But why do we need such option? Why not simply fix the code to preserve the context.

I mean unless I'm wrong the code has a bug, and we could simply fix it.

Reply:
Most Watermill Pub/Subs work over the network. You publish a message, it gets marshalled, then it's delivered to the subscriber. But obviously, the context wouldn't survive the marshaling, so you can't rely on it.

The GoChannel Pub/Sub works in memory, so it is possible to pass the context. I suppose it could be useful for passing some arbitrary data instead of marshaling it to `[]byte`, but it behaves by design the same as other Pub/Subs. Because you should be able to replace one Pub/Sub implementation with another with similar results. For example, it's handy to use GoChannel for tests, but you would use something different for production.

A scenario where passing the context may be confusing:
* An HTTP handler publishes a message passing the request's context.
* The message is delivered to a subscriber, which appends some data to the database, using the message's context.
* Meanwhile, the HTTP request is done, and the context is canceled. If you preserved the context in the message, it would also cancel the database query, which is not what you want.

In general, the publishing and subscribing sides should be considered as something separate, working concurrently. Preserving the context can make things confusing.

If you really know what you're doing, I guess it's fine to change this behavior, but you should explicitly flip the config flag then and proceed with caution (e.g., use `context.WithoutCancel` when passing it to the `Publish`).

Reply:
Thank you for your detailed reply!

Reply:
> @yashb042 It's because the context is passed to the message and is likely to be used in the command handler. For example, when executing a database query, you would pass `msg.Context()`. If the subscriber closes, the context gets canceled, which also cancels the database query.
> 
> Let me know if this makes sense?

Agreed, and the database case makes sense as well.
Though, is it there for all the cases, because as I see it, context usually is considered whenever either you do a network call (in database case you mentioned) or you call a library which has put up an exclusive check on the context.Done() field.

For our local pub-sub, I don't see any of these happening that's why a bit confused.